// mowojang/dist/src/client.js
import { isAxiosError } from "axios";
import axiosInstance from "./axiosInstance.js";
import { validateArray, validatePlayer, validateUUID } from "./validators.js";
import { undashUUID } from "./utils.js";
export default class Client {
    axios;
    validation;
    constructor(clientOptions) {
        this.axios = axiosInstance(clientOptions);
        this.validation = clientOptions?.validation ?? {};
    }
    shouldValidate(config) {
        return this.validation?.enabled || config?.validation?.enabled || false;
    }
    getValidationMinLength(config) {
        if (this.validation?.minimumUsernameLength === 1)
            return 1;
        if (this.validation?.minimumUsernameLength === 2)
            return 2;
        if (config?.validation?.minimumUsernameLength === 1)
            return 1;
        if (config?.validation?.minimumUsernameLength === 2)
            return 2;
        return undefined;
    }
    /**
     * Returns an Array of Player Profiles consisting of their Usernames and UUIDs
     *
     * Players considered "INVALID" are excluded from the results
     */
    async getProfiles(players, config) {
        try {
            if (this.shouldValidate(config) && !validateArray(players, validatePlayer, this.getValidationMinLength(config)))
                return { data: null, error: "INVALID_INPUT" };
            players = players
                .map((player) => {
                if (validateUUID(player))
                    return undashUUID(player);
                return player.toLowerCase();
            })
                .sort();
            const fetchResponse = await this.axios.post("https://mowojang.matdoes.dev/", players, {
                cache: config?.cache ?? { ttl: 15 * 60 * 1000 },
            });
            if (!Array.isArray(fetchResponse?.data))
                return { data: null, error: "UNKOWN_ERROR" };
            return {
                data: fetchResponse.data.map((player) => {
                    return {
                        UUID: player.id,
                        username: player.name,
                    };
                }),
                error: null,
            };
        }
        catch {
            return { data: null, error: "UNKOWN_ERROR" };
        }
    }
    /**
     * Returns a Player's Profile consisting of their Username and UUID
     *
     */
    async getProfile(player, config) {
        try {
            if (this.shouldValidate(config) && !validatePlayer(player, this.getValidationMinLength(config)))
                return { data: null, error: "INVALID_INPUT" };
            const fetchResponse = await this.axios.get(`https://mowojang.matdoes.dev/${player}`, {
                cache: config?.cache ?? { ttl: 15 * 60 * 1000 },
            });
            return {
                data: {
                    UUID: fetchResponse.data.id,
                    username: fetchResponse.data.name,
                },
                error: null,
            };
        }
        catch {
            return { data: null, error: "INVALID_PLAYER" };
        }
    }
    /**
     * A Simple Wrapper to Retrieve Only the UUID of a Player's Profile
     *
     */
    async getUUID(username, config) {
        const profile = await this.getProfile(username, config);
        if (profile.error)
            return null;
        return profile.data.UUID;
    }
    /**
     * A Simple Wrapper to Retrieve Only the Username of a Player's Profile
     *
     */
    async getUsername(UUID, config) {
        const profile = await this.getProfile(UUID, config);
        if (profile.error)
            return null;
        return profile.data.username;
    }
    /**
     * Returns an Array of Player Sessions consisting of their Usernames, UUIDs, Skins and Capes
     *
     * Players considered "INVALID" are excluded from the results
     */
    async getSessions(players, config) {
        try {
            if (this.shouldValidate(config) && !validateArray(players, validatePlayer, this.getValidationMinLength(config)))
                return { data: null, error: "INVALID_INPUT" };
            players = players.map((player) => {
                if (validateUUID(player))
                    return undashUUID(player);
                return player.toLowerCase();
            });
            const sessionsPromises = [];
            players.forEach((player) => {
                // @ts-ignore
                sessionsPromises.push(this.getSession(player, config));
            });
            const sessions = (await Promise.all(sessionsPromises))
                .filter((session) => session.data !== null)
                .map((session) => session.data);
            return {
                // @ts-ignore
                data: sessions,
                error: null,
            };
        }
        catch {
            return { data: null, error: "UNKOWN_ERROR" };
        }
    }
    /**
     * Returns a Player's Sessions consisting of their Username, UUID, Skin and Cape
     *
     */
    async getSession(player, config) {
        if (this.shouldValidate(config) && !validatePlayer(player, this.getValidationMinLength(config)))
            return { data: null, error: "INVALID_INPUT" };
        const UUID = await this.getUUID(player);
        if (!UUID)
            return { data: null, error: "INVALID_PLAYER" };
        return await this.axios
            .get(`https://mowojang.matdoes.dev/session/minecraft/profile/${UUID}`, {
            cache: config?.cache ?? { ttl: 15 * 60 * 1000 },
        })
            .then((fetchResponse) => {
            let textures = fetchResponse.data.properties.find((property) => property.name === "textures").value;
            let skin = null;
            let cape = null;
            if (textures) {
                textures = JSON.parse(Buffer.from(textures, "base64").toString()).textures;
                if (textures.SKIN) {
                    skin = {
                        url: textures.SKIN.url,
                        hash: textures.SKIN.url.split("/").at(-1),
                        metadata: {
                            slim: textures.SKIN?.metadata?.model === "slim",
                        },
                    };
                }
                if (textures.CAPE) {
                    cape = {
                        url: textures.CAPE.url,
                        hash: textures.CAPE.url.split("/").at(-1),
                    };
                }
            }
            return {
                data: {
                    UUID: fetchResponse.data.id,
                    username: fetchResponse.data.name,
                    skin,
                    cape,
                    actions: fetchResponse.data.profileActions,
                },
                error: null,
            };
        })
            .catch((fetchError) => {
            if (isAxiosError(fetchError)) {
                return { data: null, error: "INVALID_PLAYER" };
            }
            return { data: null, error: "UNKOWN_ERROR" };
        });
    }
    /**
     * A Simple Wrapper to Retrieve Only the Skin from a Player's Session
     *
     */
    async getSkin(player, config) {
        const session = await this.getSession(player, config);
        if (session.error)
            return null;
        return session.data.skin;
    }
    /**
     * A Simple Wrapper to Retrieve Only the Player's Skin from their Session Data Loaded into a Buffer
     *
     */
    async getSkinBuffer(player, config) {
        const session = await this.getSession(player, config);
        if (session.error || !session.data.skin)
            return null;
        const fetchResponse = await this.axios.get(session.data.skin.url, {
            responseType: "arraybuffer",
            cache: config?.cache ?? { ttl: 24 * 60 * 60 * 1000 },
        });
        return Buffer.from(fetchResponse.data, "base64");
    }
    /**
     * A Simple Wrapper to Retrieve Only the Cape Data from a Player's Session
     *
     */
    async getCape(player, config) {
        const session = await this.getSession(player, config);
        if (session.error)
            return null;
        return session.data.cape;
    }
    /**
     * A Simple Wrapper to Retrieve Only the Player's Cape from their Session Data Loaded into a Buffer
     *
     */
    async getCapeBuffer(player, config) {
        const session = await this.getSession(player, config);
        if (session.error || !session.data.cape)
            return null;
        const fetchResponse = await this.axios.get(session.data.cape.url, {
            responseType: "arraybuffer",
            cache: config?.cache ?? { ttl: 30 * 24 * 60 * 60 * 1000 },
        });
        return Buffer.from(fetchResponse.data, "base64");
    }
}
