// @hypixelic/mineflayer/src/index.js
import { Client as MowojangClient } from "mowojang";
import { getLocation, parseLocation, parseGuildChat, parseGuildEvents, parsePrivateChat, parseSkyblockCoopChat, sendGuildMessage, sendPrivateMessage, sendSkyblockCoopMessage, sendGuildOfficerMessage, toggleGuildSlowChat, muteGuildChat, muteGuildMember, unmuteGuildChat, unmuteGuildMember } from "./modules/index.js";
export const HyFlayer = (bot)=>{
  /* Structures */ bot.mowojang = new MowojangClient();
  bot.hypixel = {
    proxy: {
      ip: undefined,
      port: undefined,
      latency: undefined
    },
    location: {}
  };
  /* Functions */ bot.sendGuildMessage = (msg)=>sendGuildMessage(bot, msg);
  bot.sendGuildOfficerMessage = (msg)=>sendGuildOfficerMessage(bot, msg);
  bot.toggleGuildSlowChat = ()=>toggleGuildSlowChat(bot);
  bot.muteGuildChat = (duration)=>muteGuildChat(bot, duration);
  bot.muteGuildMember = (member, duration)=>muteGuildMember(bot, member, duration);
  bot.unmuteGuildChat = ()=>unmuteGuildChat(bot);
  bot.unmuteGuildMember = (member)=>unmuteGuildMember(bot, member);
  bot.sendPrivateMessage = (player, msg)=>sendPrivateMessage(bot, player, msg);
  bot.sendSkyblockCoopMessage = (msg)=>sendSkyblockCoopMessage(bot, msg);
  /* Proxy Parsing */ bot.once("login", ()=>{
    bot.hypixel.proxy = {
      ip: bot._client?.socket?.remoteAddress,
      port: bot._client?.socket?.remotePort,
      latency: bot?.player?.ping || undefined
    };
    setInterval(()=>{
      bot.hypixel.proxy.latency = bot?.player?.ping || undefined;
    }, 60000);
  });
  /* Location Parsing */ parseLocation(bot);
  bot.once("spawn", ()=>getLocation(bot));
  bot.on("respawn", ()=>getLocation(bot));
  /* Chat Parsers */ parsePrivateChat(bot);
  parseSkyblockCoopChat(bot);
  parseGuildChat(bot);
  parseGuildEvents(bot);
};
