// @hypixelic/mineflayer/src/modules/guild.js
export const parseGuildChat = (bot)=>{
  bot.addChatPattern("hypixel_guild_chat", /^Guild > .*:.*/);
  bot.addChatPattern("hypixel_guild_officer_chat", /^Officer > .*/);
  bot.on("chat:hypixel_guild_chat", async (msg)=>{
    var player;
    const parsedMessage = msg.toString().slice(0, msg.toString().indexOf(":")).split(" ");
    if (parsedMessage[2].includes("[") && parsedMessage[4]?.includes("[")) {
      player = parsedMessage[3];
    }
    if (!parsedMessage[2].includes("[") && !parsedMessage[4]?.includes("[")) {
      player = parsedMessage[2];
    }
    if (!parsedMessage[2].includes("[") && parsedMessage[4]?.includes("[")) {
      player = parsedMessage[2];
    }
    if (parsedMessage[2].includes("[") && !parsedMessage[4]?.includes("[")) {
      player = parsedMessage[3];
    }
    bot.emit("HYFLAYER_GUILD_CHAT", {
      UUID: await bot.mowojang.getUUID(player),
      username: player,
      message: msg.toString().slice(msg.toString().indexOf(":") + 2),
      timestamp: Math.floor(Date.now() / 1000)
    });
  });
  bot.on("chat:hypixel_guild_officer_chat", async (msg)=>{
    var player;
    const parsedMessage = msg.toString().slice(0, msg.toString().indexOf(":")).split(" ");
    if (parsedMessage[2].includes("[") && parsedMessage[4]?.includes("[")) {
      player = parsedMessage[3];
    }
    if (!parsedMessage[2].includes("[") && !parsedMessage[4]?.includes("[")) {
      player = parsedMessage[2];
    }
    if (!parsedMessage[2].includes("[") && parsedMessage[4]?.includes("[")) {
      player = parsedMessage[2];
    }
    if (parsedMessage[2].includes("[") && !parsedMessage[4]?.includes("[")) {
      player = parsedMessage[3];
    }
    bot.emit("HYFLAYER_GUILD_OFFICER_CHAT", {
      UUID: await bot.mowojang.getUUID(player),
      username: player,
      message: msg.toString().slice(msg.toString().indexOf(":") + 2),
      timestamp: Math.floor(Date.now() / 1000)
    });
  });
};
export const parseGuildEvents = (bot)=>{
  bot.addChatPattern("hypixel_guild_join", /^Guild > [^:]+ joined\.$/);
  bot.addChatPattern("hypixel_guild_leave", /^Guild > [^:]+ left\.$/);
  bot.on("chat:hypixel_guild_join", async (msg)=>{
    bot.emit("HYFLAYER_GUILD_JOIN", {
      UUID: await bot.mowojang.getUUID(msg.toString().split(" ")[2]),
      username: msg.toString().split(" ")[2],
      timestamp: Math.floor(Date.now() / 1000)
    });
  });
  bot.on("chat:hypixel_guild_leave", async (msg)=>{
    bot.emit("HYFLAYER_GUILD_LEAVE", {
      UUID: await bot.mowojang.getUUID(msg.toString().split(" ")[2]),
      username: msg.toString().split(" ")[2],
      timestamp: Math.floor(Date.now() / 1000)
    });
  });
};
export const sendGuildMessage = (bot, msg)=>{
  bot.chat(`/gc ${msg}`);
};
export const sendGuildOfficerMessage = (bot, msg)=>{
  bot.chat(`/oc ${msg}`);
};
export const toggleGuildSlowChat = (bot)=>{
  bot.chat(`/g slow`);
};
export const muteGuildChat = (bot, duration)=>{
  bot.chat(`/g mute everyone ${duration}`);
};
export const muteGuildMember = (bot, player, duration)=>{
  bot.chat(`/g mute ${player} ${duration}`);
};
export const unmuteGuildChat = (bot)=>{
  bot.chat(`/g unmute everyone`);
};
export const unmuteGuildMember = (bot, player)=>{
  bot.chat(`/g unmute ${player}`);
};
